core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -0,0 +1,501 @@
+         if (awaitResponses(rsp)) {
+            hasClear = false;
+         } else {
+            //we got an exception. go for next round
+            return;
+         }
+      }
+      try {
+         SendKeyTask task = new SendKeyTask();
+         updatedKeys.forEach(task);
+         task.await();
+      } catch (InterruptedException e) {
+         throw e;
+      } catch (Throwable t) {
+         //TODO log exception
+         log.fatal("[IRAC] Unexpected error!", t);
+      }
+   }
+
+   private XSiteResponse sendToRemoteSite(XSiteBackup backup, XSiteReplicateCommand cmd) {
+      XSiteResponse rsp = rpcManager.invokeXSite(backup, cmd);
+      takeOfflineManager.registerRequest(rsp);
+      return rsp;
+   }
+
+   private void removeKey(Object key, int segmentId, Object lockOwner, IracMetadata tombstone) {
+      //removes the key from the "updated keys map" in all owners.
+      DistributionInfo dInfo = getDistributionInfo(segmentId);
+      IracCleanupKeyCommand cmd = commandsFactory.buildIracCleanupKeyCommand(key, lockOwner, tombstone);
+      rpcManager.sendToMany(dInfo.writeOwners(), cmd, DeliverOrder.NONE);
+      cleanupKey(key, lockOwner, tombstone);
+   }
+
+   private DistributionInfo getDistributionInfoKey(Object key) {
+      return clusteringDependentLogic.getCacheTopology().getDistribution(key);
+   }
+
+   private DistributionInfo getDistributionInfo(int segmentId) {
+      return clusteringDependentLogic.getCacheTopology().getDistribution(segmentId);
+   }
+
+   private int getSegment(Object key) {
+      return clusteringDependentLogic.getCacheTopology().getSegment(key);
+   }
+
+   private boolean isWriteOwner(Object key) {
+      return getDistributionInfoKey(key).isWriteOwner();
+   }
+
+   private CompletionStage<Void> sendCommandToAllBackups(XSiteReplicateCommand command) {
+      if (command == null) {
+         return IGNORE_KEY_SEND;
+      }
+      AggregateCompletionStage<Void> collector = CompletionStages.aggregateCompletionStage();
+      for (XSiteBackup backup : asyncBackups) {
+         if (takeOfflineManager.getSiteState(backup.getSiteName()) == SiteState.OFFLINE) {
+            continue; //backup is offline
+         }
+         collector.dependsOn(sendToRemoteSite(backup, command));
+      }
+      return collector.freeze();
+   }
+
+   private XSiteReplicateCommand buildClearCommand() {
+      return commandsFactory.buildIracUpdateKeyCommand(null, null, null, null);
+   }
+
+   private XSiteReplicateCommand buildRemoveCommand(CleanupTask cleanupTask) {
+      Object key = cleanupTask.key;
+      IracMetadata metadata = iracVersionGenerator.getTombstone(key);
+      if (metadata == null) {
+         return null;
+      }
+      cleanupTask.tombstone = metadata;
+      return commandsFactory.buildIracUpdateKeyCommand(key, null, null, metadata);
+   }
+
+   private CompletionStage<InternalCacheEntry<Object, Object>> fetchEntry(Object key, int segmentId) {
```

Comment:
This method should be replaced by a reference to `EntryLoader#loadAndStoreInDataContainer` such as https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/container/entries/VersionedRepeatableReadEntry.java#L98

Reply:
It handles expiration, notifications and ordering if passivation is enabled. The data container lock is not used for this purpose anymore.

Reply:
damn... The EntryLoader is really hidden xD

Reply:
hmmm. I can change it later. The `EntryLoader` requires the `InvocationContext` and a `Command` and IRAC doesn't have that.

Reply:
The Command can just be null and the context can just be `ImmutableContext.INSTANCE`

Reply:
:+1: